### PR TITLE
Add DynamicQuantizeLinear op

### PR DESCRIPTION
**Pull request by gyulaz-htec**
Add support for [DynamicQuantizeLinear](https://github.com/onnx/onnx/blob/main/docs/Operators.md#dynamicquantizelinear) operator.
This implementation only works with static shapes due to the use of reshape. Reshape is needed to get the max and min values across the entire input tensor. Any idea on how to solve that is welcome.

Fixes: https://github.com/migraphx-benchmark/AMDMIGraphX/issues/91

### CODE REVIEW

File: test/onnx/verify_onnx.cpp [hunk 1]
```diff
@@ -351,6 +351,60 @@ TEST_CASE(depthtospace_simple_test)
     EXPECT(migraphx::verify::verify_rms_range(result_vector, gold));
 }
 
+TEST_CASE(dynamicquantizelinear_1d_test)
```
Comment (lakhinderwalia) [inline test/onnx/verify_onnx.cpp:354]: Suggestion: Please also add the first test listed in the reference:  where its zp is 153. Thanks.

https://github.com/onnx/onnx/blob/main/docs/Operators.md#dynamicquantizelinear

File: test/onnx/verify_onnx.cpp [hunk 2]
```diff
@@ -351,6 +351,60 @@ TEST_CASE(depthtospace_simple_test)
...
+    auto results = p.eval(pp);
+
+    std::vector<uint8_t> y_results;
+    results.at(0).visit([&](auto output) { y_results.assign(output.begin(), output.end()); });
+    std::vector<uint8_t> y_gold = {191, 121, 172, 96, 42, 0};
+    EXPECT(migraphx::verify::verify_rms_range(y_results, y_gold));
+
+    std::vector<float> y_scale;
+    results.at(1).visit([&](auto output) { y_scale.assign(output.begin(), output.end()); });
+    std::vector<float> y_scale_gold = {0.01568627543747425};
```
[the remark this answers is not in the document]
Comment (gyulaz-htec) [reply test/onnx/verify_onnx.cpp]: Yes it passes. Relpaced it with that value.

File: src/onnx/parse_dynamicquantizelinear.cpp [hunk 3]
```diff
@@ -0,0 +1,97 @@
...
+#include <migraphx/ranges.hpp>
+#include <migraphx/make_op.hpp>
+#include <migraphx/tune_axis.hpp>
+#include <migraphx/common.hpp>
+#include <migraphx/onnx/broadcast_qdq.hpp>
+
+namespace migraphx {
+inline namespace MIGRAPHX_INLINE_NS {
+namespace onnx {
+
```
Comment (lakhinderwalia) [inline src/onnx/parse_dynamicquantizelinear.cpp:35]: (iff you edit this file again, then..  please cut-n-paste the reference information as comments -- as in some recent operators, e.g. in qlinearconcat).

My basic design question is: the operator reference says: "A Function to fuse calculation for Scale, Zero Point and FP32->8Bit conversion of FP32 Input data.". 

But we are not fusing any calculations here.. any thoughts on it? Thanks.

File: src/onnx/parse_dynamicquantizelinear.cpp [hunk 4]
```diff
@@ -0,0 +1,97 @@
...
+        auto x_type = x_shape.type();
+        // 1. Computing y_scale
+        auto l0 = info.add_literal(migraphx::literal{migraphx::shape{x_type}, {0}});
+        // DynamicQuantizeLinear only has uint8 quantization
+        auto q_max = info.add_literal(
+            migraphx::literal{migraphx::shape{x_type}, {std::numeric_limits<uint8_t>::max()}});
+        auto q_min = info.add_literal(
+            migraphx::literal{migraphx::shape{x_type}, {std::numeric_limits<uint8_t>::min()}});
+        auto x_reshape =
```
Comment (lakhinderwalia) [inline src/onnx/parse_dynamicquantizelinear.cpp:58]: Would this step be necessary (for static shapes) if  X is 1-D? Thanks.

File: src/onnx/parse_dynamicquantizelinear.cpp (hunk 3 above)
Comment (gyulaz-htec) [reply src/onnx/parse_dynamicquantizelinear.cpp:35]: I've checked the compiled gpu version from of dynamicquantizelinear_2d_test.onnx.
From that it seems there are lot of fused instructions in kernels:
```
module: "main"
@0 = check_context::migraphx::gpu::context -> float_type, {}, {}, target_id=0
@1 = hip::hip_allocate_memory[shape=int8_type, {96}, {1},id=main:scratch] -> int8_type, {96}, {1}, target_id=0
@2 = load[offset=0,end=48](@1) -> float_type, {3, 4}, {4, 1}, target_id=0
x = @param:x -> float_type, {3, 4}, {4, 1}, target_id=0
@4 = hip::copy_to_gpu(x,@2) -> float_type, {3, 4}, {4, 1}, target_id=0
@5 = reshape_lazy[dims={12}](@4) -> float_type, {12}, {1}, target_id=0
@6 = load[offset=48,end=60](@1) -> [float_type, {1}, {1}, int64_type, {1}, {1}], target_id=0
@7 = gpu::topk[k=1,axis=0,largest=0](@5,@6) -> [float_type, {1}, {1}, int64_type, {1}, {1}], target_id=0
@8 = load[offset=64,end=68](@1) -> float_type, {1}, {1}, target_id=0
@9 = get_tuple_elem[index=0](@7) -> float_type, {1}, {1}, target_id=0
@10 = gpu::code_object[code_object=4536,symbol_name=min_kernel,global=1,local=1024,](@9,@8) -> float_type, {1}, {1}, target_id=0
@11 = load[offset=80,end=92](@1) -> [float_type, {1}, {1}, int64_type, {1}, {1}], target_id=0
@12 = gpu::topk[k=1,axis=0,largest=1](@5,@11) -> [float_type, {1}, {1}, int64_type, {1}, {1}], target_id=0
@13 = load[offset=48,end=52](@1) -> float_type, {1}, {1}, target_id=0
@14 = get_tuple_elem[index=0](@12) -> float_type, {1}, {1}, target_id=0
@15 = gpu::code_object[code_object=4872,symbol_name=max_sub_mul_kernel,global=1,local=1024,](@14,@10,@13) -> float_type, {1}, {1}, target_id=0
@16 = load[offset=80,end=81](@1) -> uint8_type, {1}, {1}, target_id=0
@17 = gpu::code_object[code_object=4976,symbol_name=neg_div_clip_nearbyint_convert_kernel,global=1,local=1024,](@10,@15,@16) -> uint8_type, {1}, {1}, target_id=0
@18 = hip::copy_from_gpu(@15) -> float_type, {1}, {1}, target_id=0
@19 = hip::copy_from_gpu(@17) -> uint8_type, {1}, {1}, target_id=0
@20 = load[offset=64,end=76](@1) -> uint8_type, {3, 4}, {4, 1}, target_id=0
@21 = multibroadcast[out_lens={3, 4},out_dyn_dims={}](@17) -> uint8_type, {3, 4}, {0, 0}, target_id=0
@22 = multibroadcast[out_lens={3, 4},out_dyn_dims={}](@15) -> float_type, {3, 4}, {0, 0}, target_id=0
@23 = gpu::code_object[code_object=5072,symbol_name=quantizelinear_kernel,global=6,local=1024,](@4,@22,@21,@20) -> uint8_type, {3, 4}, {4, 1}, target_id=0
@24 = hip::copy_from_gpu(@23) -> uint8_type, {3, 4}, {4, 1}, target_id=0
@25 = hip::sync_stream(@24,@18,@19) -> uint8_type, {3, 4}, {4, 1}, target_id=0
@26 = @return(@25,@18,@19), target_id=0
```
From my point of view, that satisfies the requirement of fusing.

Comment (gyulaz-htec) [reply src/onnx/parse_dynamicquantizelinear.cpp:35]: > (iff you edit this file again, then.. please cut-n-paste the reference information as comments -- as in some recent operators, e.g. in qlinearconcat).

Added the comments at the top of the file

File: src/onnx/parse_dynamicquantizelinear.cpp [hunk 5]
```diff
@@ -0,0 +1,151 @@
...
+        auto x_type = x_shape.type();
+        // 1. Computing y_scale
+        auto l0 = info.add_literal(migraphx::literal{migraphx::shape{x_type}, {0}});
+        // DynamicQuantizeLinear only has uint8 quantization
+        auto q_max = info.add_literal(
+            migraphx::literal{migraphx::shape{x_type}, {std::numeric_limits<uint8_t>::max()}});
+        auto q_min = info.add_literal(
+            migraphx::literal{migraphx::shape{x_type}, {std::numeric_limits<uint8_t>::min()}});
+        auto x_reshape = x;
+        if(not(x_shape.lens().size() == 1))
```
Comment (lakhinderwalia) [inline src/onnx/parse_dynamicquantizelinear.cpp:109]: Optional: Just one comparison would suffice: `if(x_shape.lens().size() != 1)`

File: src/onnx/parse_dynamicquantizelinear.cpp [hunk 6]
```diff
@@ -0,0 +1,151 @@
...
+        {
+            x_reshape = info.add_instruction(
+                migraphx::make_op("reshape", {{"dims", {x_shape.elements()}}}), x);
+        }
+        // maximum(0, max(x))
```
Comment (lakhinderwalia) [inline src/onnx/parse_dynamicquantizelinear.cpp:114]: Optional optimization. Is it possible to add a '0' to X, and bypass both the max() and min() comparisons..

File: src/onnx/parse_dynamicquantizelinear.cpp [hunk 7]
```diff
@@ -0,0 +1,151 @@
...
+        // minimum(0, min(x))
+        auto topk_min = info.add_instruction(
+            migraphx::make_op("topk", {{"axis", 0}, {"k", 1}, {"largest", false}}), x_reshape);
+        auto min_x =
+            info.add_instruction(migraphx::make_op("get_tuple_elem", {{"index", 0}}), topk_min);
+        min_x = info.add_instruction(migraphx::make_op("min"), l0, min_x);
+
+        // y_scale = (maximum(0, max(x)) - minimum(0, min(x))) / (qmax - qmin)
+        auto sub0    = info.add_instruction(migraphx::make_op("sub"), max_x, min_x);
+        auto y_scale = info.add_instruction(migraphx::make_op("div"), sub0, q_max);
```
Comment (lakhinderwalia) [inline src/onnx/parse_dynamicquantizelinear.cpp:130]: Optional. `(q_max - q_min)` instead of just `q_max` on line 130.

Comment (lakhinderwalia) [reply src/onnx/parse_dynamicquantizelinear.cpp]: Excuse me, this is a compile time step, not a new (additional) compute instruction. Hence the above suggestion. Sorry about any confusion.

File: src/onnx/parse_dynamicquantizelinear.cpp (hunk 4 above)
Comment (pfultz2) [reply src/onnx/parse_dynamicquantizelinear.cpp]: The optimizer will remove the redundant reshapes automatically so its not necessary to do this here. 

No need to revert the change either, if you already updated it, just a note for the future.

File: src/onnx/parse_dynamicquantizelinear.cpp [hunk 8]
```diff
@@ -0,0 +1,152 @@
...
+        auto x_reshape = x;
+        if(x_shape.lens().size() != 1)
+        {
+            x_reshape = info.add_instruction(
+                migraphx::make_op("reshape", {{"dims", {x_shape.elements()}}}), x);
+        }
+        x_reshape = info.add_instruction(migraphx::make_op("concat", {{"axis", 0}}), x_reshape, l0);
+        // maximum(0, max(x))
+        auto topk_max = info.add_instruction(
+            migraphx::make_op("topk", {{"axis", 0}, {"k", 1}, {"largest", true}}), x_reshape);
```
Comment (CharlieL7) [inline src/onnx/parse_dynamicquantizelinear.cpp:117]: Is it preferable to use `topk` over the equivalent of `np.max(x)`? I believe we currently parse ONNX's `Max` operator into multiple MIGX `max` operators. Probably a question for @pfultz2.

File: src/onnx/parse_dynamicquantizelinear.cpp [hunk 9]
```diff
@@ -0,0 +1,152 @@
+/*
+ * The MIT License (MIT)
+ *
+ * Copyright (c) 2015-2022 Advanced Micro Devices, Inc. All rights reserved.
```
Comment (CharlieL7) [inline src/onnx/parse_dynamicquantizelinear.cpp:4]:
```suggestion
 * Copyright (c) 2015-2023 Advanced Micro Devices, Inc. All rights reserved.
```

File: src/onnx/parse_dynamicquantizelinear.cpp [hunk 10]
```diff
@@ -111,23 +111,23 @@ struct parse_dynamicquantizelinear : op_parser<parse_dynamicquantizelinear>
...
         auto min_x =
             info.add_instruction(migraphx::make_op("get_tuple_elem", {{"index", 0}}), topk_min);
-        min_x = info.add_instruction(migraphx::make_op("min"), l0, min_x);
 
         // y_scale = (maximum(0, max(x)) - minimum(0, min(x))) / (qmax - qmin)
-        auto sub0    = info.add_instruction(migraphx::make_op("sub"), max_x, min_x);
-        auto div     = info.add_instruction(migraphx::make_op("sub"), q_max, q_min);
+        auto sub0 = info.add_instruction(migraphx::make_op("sub"), max_x, min_x);
+        // qmax - qmin is always 255
+        auto div     = q_max;
```
Comment (lakhinderwalia) [inline src/onnx/parse_dynamicquantizelinear.cpp:130]: // https://onnx.ai/onnx/operators/onnx__QuantizeLinear.html isn't any longer just for uint8. Please remove the comment on line 129. 
`auto div = q_max - q_min;` // line 130.

File: src/onnx/parse_dynamicquantizelinear.cpp (hunk 8 above)
Comment (pfultz2) [reply src/onnx/parse_dynamicquantizelinear.cpp:117]: > Is it preferable to use topk over the equivalent of np.max(x)? I

`np.max(x)` does sound like a better option. 

> I believe we currently parse ONNX's Max operator into multiple MIGX max operators.

Yea for when its variadic.

Comment (pfultz2) [reply src/onnx/parse_dynamicquantizelinear.cpp:117]: > I also looked at the choices, and [Onnx Max](https://xadupre.github.io/draft/onnx/onnx_doc_folder/onnx__Max.html) wasn't the right option. We need some kind of equivalent of std::max() & std::min(), if there is any.

Charlie is suggesting to use something equivalent to `np.max`, which would be `reduce_max` in migraphx(or [`ReduceMax`](https://xadupre.github.io/draft/onnx/onnx_doc_folder/onnx__ReduceMax.html) in onnx).

Our [`max`](https://github.com/ROCmSoftwarePlatform/AMDMIGraphX/blob/develop/src/include/migraphx/op/max.hpp#L45) and [`min`](https://github.com/ROCmSoftwarePlatform/AMDMIGraphX/blob/develop/src/include/migraphx/op/min.hpp#L45) operator already uses `std::max` and `std::min`, but this is applied per element on the tensor(similar to `np.maximum` not `np.max`).

File: src/onnx/parse_dynamicquantizelinear.cpp (hunk 6 above)
Comment (lakhinderwalia) [reply src/onnx/parse_dynamicquantizelinear.cpp:115]: Thanks.

File: src/onnx/parse_dynamicquantizelinear.cpp (hunk 10 above)
Comment (gyulaz-htec) [reply src/onnx/parse_dynamicquantizelinear.cpp:130]: The link you provided is for QuantizeLinear, DynamicQuantizeLinear only has support for uint8: https://onnx.ai/onnx/operators/onnx__DynamicQuantizeLinear.html

**Review by lakhinderwalia — APPROVED** [review]
Thank you.

**Review by CharlieL7 — APPROVED** [review]

**Pull request merged**